### PR TITLE
Inventory plugin for foreman with stub for integration tests

### PR DESCRIPTION
**Pull request by webknjaz**
##### SUMMARY

Inventory plugin for foreman

This PR includes and closes #25565. It also introduces a testing framework for it.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
foreman inventory plugin

##### ANSIBLE VERSION
```
2.5
```

### CODE REVIEW

**Comment by webknjaz** [comment]
@gundalow May I ask you to build/publish a container for the integration tests?

**Comment by mattclay** [comment]
@webknjaz The simulator needs to be split into a separate PR so it can be built first. However, given the large number of files for the simulator, I think we need to look at a few options:

1. Putting the simulator in a separate repository.
2. Reduce the number of required files.

I've already started looking at how to better manage our test containers so they no longer need to be in the ansible repository, but I haven't completed that work yet.

**Comment by webknjaz** [comment]
Simulator has been moved to --> https://github.com/ansible/foreman-test-container

File: test/runner/lib/cloud/foreman.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+
+    def cleanup(self):
+        """Clean up the resource and temporary configs files after tests."""
+        if self.container_name:
+            docker_rm(self.args, self.container_name)
+
+        super(ForemanProvider, self).cleanup()
+
+    def _setup_dynamic(self):
+        """Create a vcenter simulator using docker."""
```
Comment (mattclay) [inline test/runner/lib/cloud/foreman.py]: Leftover reference to `vcenter`.

Comment (webknjaz) [reply test/runner/lib/cloud/foreman.py]: Oops

**Comment by webknjaz** [comment]
Under Python 2.7 `deepcopy`ing big results dict triggers `MemoryError`:

https://app.shippable.com/github/ansible/ansible/runs/60057/66/console
https://app.shippable.com/github/ansible/ansible/runs/60064/66/console
```python-traceback
TASK [debug] *******************************************************************
task path: /root/ansible/test/integration/targets/foreman_inventory/test_foreman_inventory.yml:36
ERROR! Unexpected Exception, this is probably a bug: 
the full traceback was:

Traceback (most recent call last):
  File "/root/ansible/bin/ansible-playbook", line 118, in <module>
    exit_code = cli.run()
  File "/root/ansible/lib/ansible/cli/playbook.py", line 122, in run
    results = pbex.run()
  File "/root/ansible/lib/ansible/executor/playbook_executor.py", line 159, in run
    result = self._tqm.run(play=play)
  File "/root/ansible/lib/ansible/executor/task_queue_manager.py", line 290, in run
    play_return = strategy.run(iterator, play_context)
  File "/root/ansible/lib/ansible/plugins/strategy/linear.py", line 291, in run
    results += self._wait_on_pending_results(iterator)
  File "/root/ansible/lib/ansible/plugins/strategy/__init__.py", line 680, in _wait_on_pending_results
    results = self._process_pending_results(iterator)
  File "/root/ansible/lib/ansible/plugins/strategy/__init__.py", line 117, in inner
    results = func(self, iterator, one_pass=one_pass, max_passes=max_passes)
  File "/root/ansible/lib/ansible/plugins/strategy/__init__.py", line 642, in _process_pending_results
    self._tqm.send_callback('v2_runner_on_ok', task_result)
  File "/root/ansible/lib/ansible/executor/task_queue_manager.py", line 367, in send_callback
    new_args.append(arg.clean_copy())
  File "/root/ansible/lib/ansible/executor/task_result.py", line 120, in clean_copy
    result._result = deepcopy(self._result)
  File "/usr/lib/python2.7/copy.py", line 163, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python2.7/copy.py", line 257, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python2.7/copy.py", line 163, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python2.7/copy.py", line 257, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python2.7/copy.py", line 163, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python2.7/copy.py", line 257, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python2.7/copy.py", line 163, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python2.7/copy.py", line 257, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python2.7/copy.py", line 163, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python2.7/copy.py", line 257, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/usr/lib/python2.7/copy.py", line 163, in deepcopy
    y = copier(x, memo)
  File "/usr/lib/python2.7/copy.py", line 230, in _deepcopy_list
    y.append(deepcopy(a, memo))
  File "/usr/lib/python2.7/copy.py", line 190, in deepcopy
    y = _reconstruct(x, rv, 1, memo)
  File "/usr/lib/python2.7/copy.py", line 330, in _reconstruct
    memo[id(x)] = y
MemoryError
```

/cc: @jctanner @bcoca

File: test/integration/targets/foreman_inventory/test_foreman_inventory.yml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  - name: Smoke test HTTP response from Foreman stub
+    uri:
+      url: "{{ foreman_stub_heartbeat_uri }}"
+      return_content: yes
+    register: heartbeat_resp
+    failed_when: >
+        heartbeat_resp.json.status != 'ok' or heartbeat_resp.json.response != 'pong'
+
+  #### Testing start
+  - debug:
```
Comment (jctanner) [inline test/integration/targets/foreman_inventory/test_foreman_inventory.yml]: These aren't tests, they are just prints. Use the assert module to check for the existence of some keys and their values.

**Review by jctanner — DISMISSED** [review]

**Comment by mattclay** [comment]
In addition to the sanity test failure already noted by ansibot, there are integration test failures:

https://app.shippable.com/github/ansible/ansible/runs/60699/66/tests

**Comment by webknjaz** [comment]
@mattclay that was smth flaky and unrelated to my changes.

**Pull request merged**